Spanner's `ClientSideStatements.json`

Hello esteemed Google Cloud Platform devs,

I was trying to use this package with Spanner and Micronaut, but caught the following error:
```
 __  __ _                                  _   
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_ 
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_ 
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
  Micronaut (v3.0.0)

13:40:56.763 [main] INFO  i.m.context.DefaultBeanContext - Reading Startup environment from bootstrap.yml
13:40:56.768 [main] INFO  i.m.g.c.GoogleCredentialsFactory - Google Credentials from Application Default Credentials
13:40:57.413 [main] INFO  i.m.d.c.c.DistributedPropertySourceLocator - Resolved 1 configuration sources from client: compositeConfigurationClient(GCP Secret Manager Config Client)
13:40:57.429 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
13:40:57.430 [main] ERROR io.micronaut.runtime.Micronaut - Error starting Micronaut server: Bean definition [org.hibernate.SessionFactory] could not be loaded: Error instantiating bean of type  [org.hibernate.boot.registry.StandardServiceRegistry]

Path Taken: SessionFactory.hibernateSessionFactory([SessionFactoryBuilder sessionFactoryBuilder]) --> SessionFactoryBuilder.hibernateSessionFactoryBuilder([MetadataSources metadataSources],ValidatorFactory validatorFactory,Interceptor hibernateInterceptor) --> MetadataSources.hibernateMetadataSources(JpaConfiguration jpaConfiguration,[StandardServiceRegistry standardServiceRegistry]) --> StandardServiceRegistry.hibernateStandardServiceRegistry(String dataSourceName,[DataSource dataSource])
io.micronaut.context.exceptions.BeanInstantiationException: Bean definition [org.hibernate.SessionFactory] could not be loaded: Error instantiating bean of type  [org.hibernate.boot.registry.StandardServiceRegistry]

Path Taken: SessionFactory.hibernateSessionFactory([SessionFactoryBuilder sessionFactoryBuilder]) --> SessionFactoryBuilder.hibernateSessionFactoryBuilder([MetadataSources metadataSources],ValidatorFactory validatorFactory,Interceptor hibernateInterceptor) --> MetadataSources.hibernateMetadataSources(JpaConfiguration jpaConfiguration,[StandardServiceRegistry standardServiceRegistry]) --> StandardServiceRegistry.hibernateStandardServiceRegistry(String dataSourceName,[DataSource dataSource])
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1840)
	at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:235)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:3362)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:243)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:181)
	at io.micronaut.runtime.Micronaut.start(Micronaut.java:71)
	at co.cookies.odata.ApplicationKt.main(Application.kt:34)
Caused by: io.micronaut.context.exceptions.BeanInstantiationException: Error instantiating bean of type  [org.hibernate.boot.registry.StandardServiceRegistry]

Path Taken: SessionFactory.hibernateSessionFactory([SessionFactoryBuilder sessionFactoryBuilder]) --> SessionFactoryBuilder.hibernateSessionFactoryBuilder([MetadataSources metadataSources],ValidatorFactory validatorFactory,Interceptor hibernateInterceptor) --> MetadataSources.hibernateMetadataSources(JpaConfiguration jpaConfiguration,[StandardServiceRegistry standardServiceRegistry]) --> StandardServiceRegistry.hibernateStandardServiceRegistry(String dataSourceName,[DataSource dataSource])
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2265)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2723)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2685)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1555)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1052)
	at io.micronaut.configuration.hibernate.jpa.$EntityManagerFactoryBean$HibernateStandardServiceRegistry0Definition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:132)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2723)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2685)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1555)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1052)
	at io.micronaut.configuration.hibernate.jpa.$EntityManagerFactoryBean$HibernateMetadataSources1Definition.doBuild(Unknown Source)
	at io.micronaut.context.AbstractParametrizedBeanDefinition.build(AbstractParametrizedBeanDefinition.java:118)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:132)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2723)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2685)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1555)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1052)
	at io.micronaut.configuration.hibernate.jpa.$EntityManagerFactoryBean$HibernateSessionFactoryBuilder2Definition.build(Unknown Source)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:136)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.getBeanForDefinition(DefaultBeanContext.java:2723)
	at io.micronaut.context.DefaultBeanContext.getBeanInternal(DefaultBeanContext.java:2685)
	at io.micronaut.context.DefaultBeanContext.getBean(DefaultBeanContext.java:1555)
	at io.micronaut.context.AbstractBeanDefinition.getBeanForConstructorArgument(AbstractBeanDefinition.java:1052)
	at io.micronaut.configuration.hibernate.jpa.$EntityManagerFactoryBean$HibernateSessionFactory3Definition.build(Unknown Source)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:136)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingletonInternal(DefaultBeanContext.java:3193)
	at io.micronaut.context.DefaultBeanContext.createAndRegisterSingleton(DefaultBeanContext.java:3179)
	at io.micronaut.context.DefaultBeanContext.loadContextScopeBean(DefaultBeanContext.java:2555)
	at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1834)
	... 6 common frames omitted
Caused by: java.lang.ExceptionInInitializerError: null
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:315)
	at com.google.cloud.spanner.connection.StatementParser.<init>(StatementParser.java:224)
	at com.google.cloud.spanner.connection.StatementParser.<clinit>(StatementParser.java:45)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:375)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:295)
	at com.google.cloud.spanner.connection.ConnectionImpl.<init>(ConnectionImpl.java:84)
	at com.google.cloud.spanner.connection.ConnectionOptions.getConnection(ConnectionOptions.java:784)
	at com.google.cloud.spanner.jdbc.AbstractJdbcConnection.<init>(AbstractJdbcConnection.java:61)
	at com.google.cloud.spanner.jdbc.JdbcConnection.<init>(JdbcConnection.java:58)
	at com.google.cloud.spanner.jdbc.JdbcDriver.connect(JdbcDriver.java:194)
	at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:121)
	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:364)
	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:476)
	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
	at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
	at io.micronaut.configuration.jdbc.hikari.HikariUrlDataSource.<init>(HikariUrlDataSource.java:35)
	at io.micronaut.configuration.jdbc.hikari.DatasourceFactory.dataSource(DatasourceFactory.java:66)
	at io.micronaut.configuration.jdbc.hikari.$DatasourceFactory$DataSource0Definition.build(Unknown Source)
	at io.micronaut.context.BeanDefinitionDelegate.build(BeanDefinitionDelegate.java:136)
	at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2238)
	... 48 common frames omitted
Caused by: java.lang.NullPointerException: null
	at java.io.Reader.<init>(Reader.java:167)
	at java.io.InputStreamReader.<init>(InputStreamReader.java:72)
	at com.google.cloud.spanner.connection.ClientSideStatements.importStatements(ClientSideStatements.java:36)
	at com.google.cloud.spanner.connection.ClientSideStatements.<clinit>(ClientSideStatements.java:27)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.invokeClassInitializer(ClassInitializationInfo.java:375)
	at com.oracle.svm.core.classinitialization.ClassInitializationInfo.initialize(ClassInitializationInfo.java:295)
	... 69 common frames omitted
```

It seems in Spanner Java's [`ClientSideStatements.java`](https://github.com/googleapis/java-spanner/blob/865bac1a976d16e20e0be0eed11594c7c2a7d721/google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ClientSideStatements.java#L25-L36), it loads a file ([`ClientSideStatements.json`](https://github.com/googleapis/java-spanner/blob/master/google-cloud-spanner/src/main/resources/com/google/cloud/spanner/connection/ClientSideStatements.json)) which doesn't make it into the image.

Over in [`SpannerFeature.java`](), I don't see any resource declarations like are present in, say, [`GoogleJsonClientFeature.java`](https://github.com/GoogleCloudPlatform/google-cloud-graalvm-support/blob/master/google-cloud-graalvm-support/src/main/java/com/google/cloud/graalvm/features/core/GoogleJsonClientFeature.java#L63-L70).

It seems like these added lines in [`SpannerFeature`](https://github.com/GoogleCloudPlatform/google-cloud-graalvm-support/blob/338e6c1797fff35946b13175c1cf6501de1c77a2/google-cloud-graalvm-support/src/main/java/com/google/cloud/graalvm/features/SpannerFeature.java#L35-L42) might fix it:
```diff
+ import com.oracle.svm.core.configure.ResourcesRegistry;
+ import org.graalvm.nativeimage.ImageSingletons;
// ...
    Class<?> spannerClass = access.findClassByName(SPANNER_CLASS);
    if (spannerClass != null) {
+    ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+    resourcesRegistry.addResources(
+        "\\Qcom/google/cloud/spanner/connection/ClientSideStatements.json\\E");
+    resourcesRegistry.addResources(
+        "\\Qcom/google/cloud/spanner/spi/v1/grpc-gcp-apiconfig.json\\E");
    }
```

Cloud BOM Version: `22.0.0`
Spanner Version: `6.12.2`
Micronaut Version: `3.0.0`
GraalVM Version: `21.2.0`
This lib: `0.6.0`

(All latest at the time of this filing)

---

For anyone coming across this later, we worked around this by inlining Spanner's two resource files -- [`ClientSideStatements.json`](https://raw.githubusercontent.com/googleapis/java-spanner/master/google-cloud-spanner/src/main/resources/com/google/cloud/spanner/connection/ClientSideStatements.json) and [`grpc-gcp-apiconfig.json`](https://raw.githubusercontent.com/googleapis/java-spanner/master/google-cloud-spanner/src/main/resources/com/google/cloud/spanner/spi/v1/grpc-gcp-apiconfig.json) in our own `src/main/resources` root.

--- a/google-cloud-graalvm-support/src/main/java/com/google/cloud/graalvm/features/SpannerFeature.java
+++ b/google-cloud-graalvm-support/src/main/java/com/google/cloud/graalvm/features/SpannerFeature.java
@@ -20,6 +20,8 @@ import static com.google.cloud.graalvm.features.NativeImageUtils.registerClassFo
 import static com.google.cloud.graalvm.features.NativeImageUtils.registerClassHierarchyForReflection;
 
 import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.core.configure.ResourcesRegistry;
+import org.graalvm.nativeimage.ImageSingletons;
 import org.graalvm.nativeimage.hosted.Feature;
 
 /**
@@ -40,6 +42,13 @@ public class SpannerFeature implements Feature {
           access, "com.google.spanner.admin.instance.v1.Instance");
       registerClassForReflection(
           access, "com.google.spanner.admin.database.v1.RestoreInfo");
+
+      // Resources
+      ResourcesRegistry resourcesRegistry = ImageSingletons.lookup(ResourcesRegistry.class);
+      resourcesRegistry.addResources(
+          "\\Qcom/google/cloud/spanner/connection/ClientSideStatements.json\\E");
+      resourcesRegistry.addResources(
+          "\\Qcom/google/cloud/spanner/spi/v1/grpc-gcp-apiconfig.json\\E");
     }
   }
 }

